include/ruby/buffer.h

```diff
@@ -0,0 +1,108 @@
+#ifndef RUBY_BUFFER_H
+#define RUBY_BUFFER_H 1
+/**
+ * @file
+ * @author     Ruby developers <ruby-core@ruby-lang.org>
+ * @copyright  This  file  is   a  part  of  the   programming  language  Ruby.
+ *             Permission  is hereby  granted,  to  either redistribute  and/or
+ *             modify this file, provided that  the conditions mentioned in the
+ *             file COPYING are met.  Consult the file for details.
+ * @brief      Buffer Protocol.
+ */
+
+#include "ruby/internal/dllexport.h"
+#include "ruby/internal/value.h"
+
+enum ruby_buffer_flags {
+    RUBY_BUFFER_SIMPLE            = 0,
+    RUBY_BUFFER_WRITABLE          = (1<<0),
+    RUBY_BUFFER_FORMAT            = (1<<1),
+    RUBY_BUFFER_MULTI_DIMENSIONAL = (1<<2),
+    RUBY_BUFFER_STRIDES           = (1<<3) | RUBY_BUFFER_MULTI_DIMENSIONAL,
+    RUBY_BUFFER_ROW_MAJOR         = (1<<4) | RUBY_BUFFER_STRIDES,
+    RUBY_BUFFER_COLUMN_MAJOR      = (1<<5) | RUBY_BUFFER_STRIDES,
+    RUBY_BUFFER_ANY_CONTIGUOUS    = RUBY_BUFFER_ROW_MAJOR | RUBY_BUFFER_COLUMN_MAJOR,
+    RUBY_BUFFER_INDIRECT          = (1<<6) | RUBY_BUFFER_STRIDES,
+};
+
+typedef struct {
+    /* The original object that have the memory exported via this buffer.
+     * The consumer of this buffer has the responsibility to call rb_gc_mark
+     * for preventing this obj collected by GC.  */
+    VALUE obj;
+
+    /* The pointer to the exported memory. */
+    void *data;
+
+    /* The number of bytes in data. */
+    ssize_t len;
+
+    /* 1 for readonly memory, 0 for writable memory. */
+    int readonly;
```

Comment:
```suggestion
    int read_only;
```

Reply:
I use `readonly` instead of `read_only` to follow the existence symbols including `readonly` such as `rb_define_readonly_variable`.